lib/compact.js

```diff
@@ -308,40 +313,100 @@ api.compact = ({
           }
         }
 
-        // handle simple @graph
-        if(isSimpleGraph && !container.includes('@graph')) {
-          // wrap using @graph alias
-          compactedItem = {
-            [api.compactIri({activeCtx, iri: '@graph'})]: compactedItem
-          };
-
-          // include @index from expanded @graph, if any
-          if('@index' in expandedItem) {
-            compactedItem[api.compactIri({activeCtx, iri: '@index'})] =
-              expandedItem['@index'];
-          }
-        }
+        // Graph object compaction cases
+        if(isGraph) {
+          if(container.includes('@graph') && container.includes('@id')) {
+            // container includes @graph and @id
+            // get or create the map object
+            let mapObject;
+            if(itemActiveProperty in rval) {
+              mapObject = rval[itemActiveProperty];
+            } else {
+              rval[itemActiveProperty] = mapObject = {};
+            }
+            // add compactedItem to map, using value of `@id` or a new blank node identifier
+            _addValue(
+              mapObject, (expandedItem['@id'] || issuer.getId(null)), compactedItem,
+              {propertyIsArray: (!options.compactArrays || container.includes('@set'))});
+          } else if(container.includes('@graph') && container.includes('@index') && _isSimpleGraph(expandedItem)) {
+            // container includes @graph and @index and item is a simple graph
+            // get or create the map object
+            let mapObject;
+            if(itemActiveProperty in rval) {
+              mapObject = rval[itemActiveProperty];
+            } else {
+              rval[itemActiveProperty] = mapObject = {};
+            }
+            // add compactedItem to map, using value of `@index` or `@none`
+            _addValue(
+              mapObject, (expandedItem['@index'] || '@none'), compactedItem,
+              {propertyIsArray: (!options.compactArrays || container.includes('@set'))});
```

Comment:
The code in these first two cases is really similar and could potentially be combined -- but that can wait until we do some clean up later. Just making a note of it.

Reply:
This better?

Reply:
I meant the code is the same for the cases where `@index` and where `@id` are present other than the second parameter to `_addValue`.